src/components/AdvertisingProvider.js

```diff
@@ -7,43 +7,73 @@ import AdvertisingContext from '../AdvertisingContext';
 export default class AdvertisingProvider extends Component {
     constructor(props) {
         super(props);
+        this.initialize();
 
-        const { config, plugins } = this.props;
-        this.advertising = new Advertising(config, plugins);
-        this.activate = this.advertising.activate.bind(this.advertising);
-        this.needTearDown = false;
+        const activate = this.advertising.activate.bind(this.advertising);
+        this.state = {
+            activate
+        };
     }
 
-    async setupAdvertising() {
-        await this.advertising.setup();
-        // teardown and setup should run in pair
-        this.needTearDown = true;
+    componentDidMount() {
+        if (this.advertising.isConfigReady() && this.props.active) {
+            this.setupAdvertising();
+        }
     }
 
-    componentDidUpdate() {
+    componentDidUpdate(prevProps) {
         const { config, active } = this.props;
+        const isConfigReady = this.advertising.isConfigReady();
 
         // activate advertising when the config changes from `undefined`
-        if (!this.advertising.isConfigReady() && config && active) {
+        if (!isConfigReady && config && active) {
             this.advertising.setConfig(config);
             this.setupAdvertising();
-        }
-    }
+        } else if (isConfigReady && prevProps.config !== config) {
```

Comment:
please do a deep comparison between the two config objects here, we should avoid accidental ad refresh

Reply:
Even if I add a `return` statement here, it still need to wrap the remaining condition. Since `!isConfigReady && config && active` is not negation of `isConfigReady && prevProps.config !== config`